### PR TITLE
Don't allow Coinbase only blocks

### LINKED ISSUE #1601: Coinbase only TX sets should not validate
Currently there are no rules against network creating blocks consisting of Coinbase TXs only. This is probably not ideal. We should both not create such nominations and invalidate any nominations/blocks that only include Coinbase TXs.

### PR DESCRIPTION
**Pull request by omerfirmak**
Fixes #1601

### CODE REVIEW

File: source/agora/node/Ledger.d [hunk 1]
```diff
@@ -1982,3 +1990,43 @@ unittest
             mpv_stake.output.address).array.length == 0);
     }
 }
+
+// Coinbase only ConsensusData and blocks
```
Comment (linked0) [inline source/agora/node/Ledger.d]: Please leave a little more detailed description of what the situation is and what the expectation is.

File: source/agora/test/Fee.d [hunk 2]
```diff
@@ -219,3 +219,37 @@ unittest
     auto bad_node = nodes[0];
     assert(bad_node.getBlockHeight() == Height(0));
 }
+
+unittest
```
Comment (linked0) [inline source/agora/test/Fee.d:225]: Please leave the description.

File: source/agora/node/Ledger.d [hunk 3]
```diff
@@ -1982,3 +1990,43 @@ unittest
...
+    import agora.consensus.data.genesis.Test;
+    import std;
+    ConsensusConfig config = { validator_cycle: 20, payout_period: 1 };
+    auto params = new immutable(ConsensusParams)(GenesisBlock,
+        CommonsBudget.address, config);
+
+    const(Block)[] blocks = [ GenesisBlock ];
+    scope ledger = new TestLedger(WK.Keys.NODE2, blocks, params);
+
+    Amount per_tx_fee = Amount.UnitPerCoin;
```
Comment (linked0) [inline source/agora/node/Ledger.d]: The `per_tx_fee` doesn't need to be introduced.

**Review by linked0 — APPROVED** [review]
Good work.
You should add comments to the commit.

**Review by kchulj — APPROVED** [review]
The code LGTM.

File: source/agora/consensus/validation/Block.d [hunk 4]
```diff
@@ -137,13 +137,18 @@ public string isInvalidReason (const ref Block block, Height prev_height,
...
     foreach (const ref tx; block.txs)
     {
+        only_coinbase &= tx.type == TxType.Coinbase;
         if (auto fail_reason = VTx.isInvalidReason(tx, findUTXO,
             block.header.height, checkFee))
             return fail_reason;
     }
 
+    if (only_coinbase)
+        return "Block: Coinbase only block";
```
Comment (Geod24) [inline source/agora/consensus/validation/Block.d]:
```suggestion
        return "Block: Must contain other transactions than Coinbase";
```

We should always aim to tell what is wrong, not just the state. As a user, would you rather see "Error: Block has 1337 transactions" or "Error: Number of transaction per block must be a power of 2, not 1337" ?

File: source/agora/node/Ledger.d [hunk 5]
```diff
@@ -718,6 +720,7 @@ public class Ledger
         NoTransactions = "Transaction set doesn't contain any transactions",
         NotEnoughValidators = "Enrollment: Insufficient number of active validators",
         MayBeValid = "May be valid",
+        OnlyCoinbaseTX = "Transaction set only includes Coinbase transactions",
```
Comment (Geod24) [inline source/agora/node/Ledger.d]:
```suggestion
        OnlyCoinbaseTX = "Transaction set only includes a Coinbase transaction",
```

We can only have one, right ?

File: source/agora/test/Fee.d [hunk 6]
```diff
@@ -219,3 +219,39 @@ unittest
...
+
+    Transaction[] txs;
+    // create enough tx's for a single block
+    txs = blocks[$ - 1].spendable().map!(txb => txb
+        .deduct(Amount.UnitPerCoin).sign()).array();
+    // Send a single TX with fees to a node
+    node_1.putTransaction(txs[0]);
+    network.expectBlock(Height(1), blocks[0].header);
+    // The fees from the last block should not trigger a new block creation
+    assertThrown!AssertError(network.expectBlock(Height(2), blocks[0].header));
```
Comment (Geod24) [inline source/agora/test/Fee.d]: We should *never* test for `AssertError`. The way to do it would be to instrument a validator.

**Comment by omerfirmak** [comment]
Rebased and all comments addressed.

**Pull request merged**